src/Symfony/Component/DependencyInjection/Compiler/AliasDeprecatedPublicServicesPass.php

```diff
@@ -0,0 +1,73 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Exception\InvalidArgumentException;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+use Symfony\Component\DependencyInjection\Reference;
+
+final class AliasDeprecatedPublicServicesPass extends AbstractRecursivePass
+{
+    private $tagName;
+
+    private $aliases = [];
+
+    public function __construct(string $tagName = 'container.private')
+    {
+        $this->tagName = $tagName;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function processValue($value, bool $isRoot = false)
+    {
+        if ($value instanceof Reference && isset($this->aliases[$id = (string) $value])) {
+            return new Reference($this->aliases[$id]);
```

Comment:
this must preserve the other configuration settings of the reference, rather than always turning them into hard references.

Reply:
You mean `new Reference($this->aliases[$id], $value->getInvalidBehavior())`?